4 add sampling functions

crispy-wonton opened the pull request:
Fixes #4 

# Description
Changes:
- add `sampling.py` with sampling functions to subset EPC dataset by LSOA. 2 LSOAs are sampled per nation: one with max UPRN count and one with min UPRN count.
- add new feature processing functions and move existing ones from `getters.get_datasets.py` to `pipeline.prepare_features.py`
- rename `enhance_epc_functions.py` to `enhance_epc.py`

# Instructions for Reviewer

In order to test the code in this PR you need to run the following script in a notebook:
```
import polars as pl
from asf_heat_pump_suitability.pipeline import run_script, sampling

epc_path = "s3://asf-daps/lakehouse/processed/epc/deduplicated/processed_dedupl-0.parquet"

test = run_script.main(epc_path=epc_path)
sample = sampling.filter_df_area_sample_epc(test)
```

Please could you:
- check for errors/bugs issues
- check output works and is sensible. Output should include data from 6 LSOAs: 2 for each nation with min and max UPRN counts.

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

crispy-wonton committed:
separate feature processing functions from getters
move feature processing functions from get_datasets.py to prepare_features.py

crispy-wonton committed:
rename enhance_epc_functions.py to enhance_epc.py
- remove function to standardise postcode col for EPC because it duplicates postcode standardisation col in prepare_features.py

crispy-wonton committed:
refactor run script to use renamed enhance_epc.py

crispy-wonton committed:
add sampling.py file with functions to subset EPC data by LSOA for testing

lizgzil commented on asf_heat_pump_suitability/pipeline/prepare_features.py:
```diff
@@ -0,0 +1,72 @@
+        pl.DataFrame: dataset with new `lsoa` column
+
+    """
+    df = df.with_columns(pl.col(["lsoa11", "lsoa21"]).replace("", None))
+    df = df.with_columns(pl.col("lsoa21").fill_null(pl.col("lsoa11")).alias("lsoa"))
```
Might we want the same done with MSOA?

crispy-wonton replied:
updated function with new `area_type` param to allow inputting different output area types.

lizgzil commented on asf_heat_pump_suitability/pipeline/sampling.py:
```diff
@@ -0,0 +1,71 @@
+    return [
+        df.filter(pl.col("country_code") == cc).sample(1, seed=8).to_dicts()[0]
+        for cc in country_codes
+    ]
```
I would turn this into a script that can be run with

```
if __name__ == '__main__':
    epc_path = "s3://asf-daps/lakehouse/processed/epc/deduplicated/processed_dedupl-0.parquet"

    test = run_script.main(epc_path=epc_path, save_output=None)
    sample = sampling.filter_df_area_sample_epc(test)
    
    # Save sample on S3 somewhere
    ...
```

lizgzil replied:
I guess you'll need `from asf_heat_pump_suitability.pipeline import run_script` too

lizgzil reviewed (commented):
This looks great and the command given ran and produced a sample that I think looks as it should! I had to turn it into a pandas df to check it though since I don't know any of the polars commands 😆 

I guess in terms of turning this repo into a clearly documented and reproducible pipeline, I would add the following.
- turn the sampling.py into a script to be run (rather than running the commands in a notebook)
- save the sample output (seems unnecessary to have to resample every time when we could just be reading this dataset)
- add to a pipeline README e.g. "To create and save a sample run `python asf_heat_pump_suitability/pipeline/sampling.py`" (I don't really think it's necessary at this stage to spend time adding arguments since this sampling is such an ad hoc part of the full pipeline)

Another point was about MSOA. Since one of the datasets you found was per MSOA - should we perhaps save a sample per MSOA too? I guess you've made this so easy to create with `sampling.filter_df_area_sample_epc(test, area='msoa')` which is great!

crispy-wonton committed:
add pipeline.README.md with instructions to run sampling.py script

crispy-wonton committed:
change _clean_col_lsoa to _clean_col_output_area
- add new param to function to allow specification of output area type
- refactor remaining code which uses function

crispy-wonton committed:
add run script to sampling.py
- run script generates EPC sample and saves to S3

lizgzil commented on asf_heat_pump_suitability/pipeline/README.md:
```diff
@@ -0,0 +1,17 @@
+# asf_heat_pump_suitability.pipeline
```
lovely :)

crispy-wonton replied:
@lizgzil I've added a new section to this README with instructions for running `run_script.py`

lizgzil reviewed (dismissed):
this looks great and then script ran. Thanks for making the changes :)

crispy-wonton committed:
add more cols to EPC dataset config

crispy-wonton committed:
add run_script.py instructions to pipeline.README
